p2p/stream.go

```diff
@@ -0,0 +1,84 @@
+package p2p
+
+import (
+	"io"
+	"sync"
+
+	manet "gx/ipfs/QmNqRnejxJxjRroz7buhrjfU8i3yNBLa81hFtmf2pXEffN/go-multiaddr-net"
+	ma "gx/ipfs/QmUxSEGbv2nmYNnfXi7839wwQqTN3kwQeUxe8dTjZWZs7J/go-multiaddr"
+	net "gx/ipfs/QmXdgNhVEgjLxjUoMs5ViQL7pboAt3Y7V7eGHRiE4qrmTE/go-libp2p-net"
+	"gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+)
+
+// Stream holds information on active incoming and outgoing p2p streams.
+type Stream struct {
+	id uint64
+
+	Protocol protocol.ID
+
+	OriginAddr ma.Multiaddr
+	TargetAddr ma.Multiaddr
+
+	Local  manet.Conn
+	Remote net.Stream
+
+	Registry *StreamRegistry
+}
+
+// Close closes stream endpoints and deregisters it
+func (s *Stream) Close() error {
+	s.Local.Close()
+	s.Remote.Close()
+	s.Registry.Deregister(s.id)
+	return nil
+}
+
+// Reset closes stream endpoints and deregisters it
+func (s *Stream) Reset() error {
+	s.Local.Close()
+	s.Remote.Reset()
+	s.Registry.Deregister(s.id)
+	return nil
+}
+
+func (s *Stream) startStreaming() {
+	go func() {
+		io.Copy(s.Local, s.Remote)
+		s.Reset()
```

Comment:
This should only reset on error.

Ideally, we'd support half-closed streams but I don't believe we can feasibly do that.